test/jdk/java/util/zip/Available.java

```diff
@@ -22,69 +22,162 @@
+        Files.deleteIfExists(zip);
+    }
 
-        // test ZipInputStream
-        try (FileInputStream fis = new FileInputStream(f);
-             ZipInputStream z = new ZipInputStream(fis))
-        {
+    /**
+     * Verify that ZipInputStream.available() returns 0 after EOF or
+     * closeEntry, otherwise 1, as specified in the API description.
+     * This tests 4028605 4109069 4234207
+     * @throws IOException if an unexpected error occurs
+     */
+    @Test
+    public void testZipInputStream() throws IOException {
+        try (InputStream in = Files.newInputStream(zip)) {
+            ZipInputStream z = new ZipInputStream(in);
             z.getNextEntry();
-            tryAvail(z);
+            assertEquals(1, z.available());
+            z.read();
+            assertEquals(1, z.available());
+            z.transferTo(OutputStream.nullOutputStream());
+            assertEquals(0, z.available(),
+                    "ZipInputStream.available() should return 0 after EOF");
+
+            z.close();
+            assertThrows(IOException.class, () -> z.available(),
+                    "Expected an IOException when calling available on a closed stream");
         }
 
-        // test InflaterInputStream
-        try (ZipFile zfile = new ZipFile(f)) {
-            tryAvail(zfile.getInputStream(zfile.getEntry("Available.java")));
+        try (InputStream in = Files.newInputStream(zip);
+             ZipInputStream z = new ZipInputStream(in)) {
+            z.getNextEntry();
+            z.closeEntry();
+            assertEquals(0, z.available(),
+                    "ZipInputStream.available() should return 0 after closeEntry");
         }
     }
 
-    static void tryAvail(InputStream in) throws Exception {
-        byte[] buf = new byte[1024];
-        int n;
+    /**
+     * Verify that ZipFileInputStream.available() returns
+     * the number of remaining uncompressed bytes.
+     *
+     * This verifies unspecified, but long-standing behavior. See 4401122.
+     *
+     * @throws IOException if an unexpected error occurs
+     */
+    @Test
+    public void testZipFileInputStream() throws IOException {
+        try (ZipFile zfile = new ZipFile(zip.toFile())) {
+            assertRemainingUncompressedBytes(zfile, "stored.txt");
+        }
+    }
 
-        while ((n = in.read(buf)) != -1);
-        if (in.available() != 0) {
-            throw new Exception("available should return 0 after EOF");
+    /**
+     * Verify that ZipFileInflaterInputStream.available() returns
+     * the number of remaining uncompressed bytes.
+     *
+     * This verifies unspecified, but long-standing behavior. See 4401122.
+     *
+     * @throws IOException if an unexpected error occurs
+     */
+    @Test
+    public void testZipFileInflaterInputStream() throws IOException {
+        try (ZipFile zfile = new ZipFile(zip.toFile())) {
+            assertRemainingUncompressedBytes(zfile, "deflated.txt");
         }
     }
 
```

Comment:
Could we collapse the two above tests  via

```
@ParameterizedTest
@ValueSource(strings = { "stored.txt", "delated.txt" })
void testAvailbleRemainingBytes(iString zipEntry) {
    try (ZipFile zfile = new ZipFile(zip.toFile())) {
            assertRemainingUncompressedBytes(zfile, zipEntry);
        }
}
```

Reply:
I like it!

I collapsed `ZipFile` related tests into the method `testZipFileStreamsRemainingBytes` and inlined the `assertRemainingUncompressedBytes` method into that.

A few minor improvements to the inlined `assertRemainingUncompressedBytes` code body:

* Added a comment saying that the `InputStream` could be `ZipFileInputStream` or `ZipFileInflaterInputStream`
* Moved the "decrement by one" assertion logic into the loop, asserting decrement on each iteration 
* Cleaned up some spacing issues in the for loop